### DISCUSSION TITLE
how retrieve network information

### DISCUSSION THREAD

**Discussion by douglasjava** [hyperledger/fabric-gateway#595] [Q&A] [answered]
I'm trying to retrieve some information from the network, but the method below apparently doesn't work or I'm using it incorrectly, most methods are using the CloseableIterator

Is the form below correct? do we have an example to recover the same

new version
```java
try (var gateway = builder.connect()) {

    var network = gateway.getNetwork(CHANNEL_NAME);
                
    int idx = 0;
    while(network.getBlockEvents().hasNext()) {

        Block next = network.getBlockEvents().next();
        infos.add(InfoNetworkDTO.builder()
                .dataCount(next.getData().getDataCount())
                .dataList(next.getData().getDataList())
                .data(next.getData().getData(0))
                .build()
        );
        idx++;
    }
}
```
            
 old version
 
 ```java
Channel channel = network.getChannel();

Collection<Peer> peersOrganizationMSPIDs = channel.getPeersForOrganization(mspid);
List<PeerModel> peerModels = new ArrayList<>();

peersOrganizationMSPIDs.forEach(peer -> peerModels.add(
        new PeerModel(peer.getName(),
                peer.getUrl(),
                peer.getProperties().toString(),
                peer.getProtocol(),
                peer.getPeerEventingServiceDisconnected().toString())
        ));
```

**Comment by bestbeforetoday** [comment, marked as answer]
You have mixed two very different capabilities here. Your **old version** is accessing information obtained by the client SDK using the Fabric discovery service about the network topology. Your **new version** is using block eventing (incorrectly) to obtain committed block from the blockchain, which is useful for tasks like [building an off-chain data store](https://github.com/hyperledger/fabric-samples/tree/main/off_chain_data#readme).

Legacy SDKs needed to use the discovery service in order to identify and then connect to network peers required for transaction endorsement. The Fabric Gateway client API does not deal with service discovery at the client. Instead, the work of identifying endorsing peer is handled by the gateway service in the peer, which already holds locally all the information returned by the discovery service. The comparison of the transaction flow between legacy SDKs and using Fabric Gateway in the [full-stack-asset-transfer-guide](https://github.com/hyperledger/fabric-samples/blob/main/full-stack-asset-transfer-guide/docs/ApplicationDev/01-FabricGateway.md) sample provides a more detailed description.

In addition to the [off_chain_data](https://github.com/hyperledger/fabric-samples/tree/main/off_chain_data#readme) sample mentioned above, the [API documentation](https://hyperledger.github.io/fabric-gateway/main/api/java/org/hyperledger/fabric/client/Network.html) contains code examples demonstrating how to use block eventing correctly.
